File: problems/0010_regular-expression-matching/bottom-to-top/solution.go
func isMatch(s string, p string) bool {
    sLen, pLen := len(s), len(p)
    memo := make([][]bool, sLen + 1)
    for i := range memo {
        memo[i] = make([]bool, pLen + 1)
    }
    memo[sLen][pLen] = true
    for i := sLen; i >= 0; i-- {
        for j := pLen - 1; j >= 0; j-- {
            matchChar := (i < sLen &&
                    (p[j] == s[i] || p[j] == '.'))
            if j < pLen - 1 && p[j + 1] == '*' {
                memo[i][j] = (memo[i][j + 2] || (matchChar && memo[i + 1][j]))
            } else {
                memo[i][j] = (matchChar && memo[i + 1][j + 1])
            }
        }
    }
    return memo[0][0]
}

File: problems/0010_regular-expression-matching/bottom-to-top/Solution.py
#!/usr/bin/python3

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        sLen, pLen = len(s), len(p)
        memo = [[False for j in range(pLen + 1)] for i in range(sLen + 1)]
        memo[sLen][pLen] = True
        for i in range(sLen, -1, -1):
            for j in range(pLen - 1, -1, -1):
                matchChar = (i < sLen and \
                        (p[j] == s[i] or p[j] == '.'))
                memo[i][j] = (memo[i][j + 2] or (matchChar and memo[i + 1][j])) \
                        if j < pLen - 1 and p[j + 1] == '*' else \
                        (matchChar and memo[i + 1][j + 1])
        return memo[0][0]
